presto-orc/src/main/java/com/facebook/presto/orc/OrcSelectiveRecordReader.java

```diff
@@ -164,29 +170,95 @@ public OrcSelectiveRecordReader(
             this.constantValues[zeroBasedIndices.get(entry.getKey())] = entry.getValue();
         }
 
+        Map<Integer, Map<Subfield, TupleDomainFilter>> filtersWithNewIndices = filters
+                .entrySet()
+                .stream()
+                .collect(toImmutableMap(entry -> zeroBasedIndices.get(entry.getKey()), Map.Entry::getValue ))
+                ;
+
         // Initial order of stream readers is:
-        //  - readers with simple filters
+        //  - readers with integer equality
+        //  - readers with integer range / multivalues / inequality
+        //  - readers with filters
         //  - followed by readers for columns that provide input to filter functions
         //  - followed by readers for columns that doesn't have any filtering
-        streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), columnsWithFilters, filterFunctionInputs);
+        streamReaderOrder = orderStreamReaders(columnTypes.keySet().stream().filter(index -> this.constantValues[index] == null).collect(toImmutableSet()), filtersWithNewIndices, filterFunctionInputs);
     }
 
-    private static int[] orderStreamReaders(Collection<Integer> columnIndices, Set<Integer> columnsWithFilters, Set<Integer> filterFunctionInputs)
+    private static int scoreFilter(Map<Subfield, TupleDomainFilter> filters){
+        final int lowestScore = 10;
+        final int lowScore = 50;
+        final int mediumScore = 100;
+        final int maxScore = 1000;
+
+        checkArgument(!filters.isEmpty());
+
+        if (filters.size() > 1){
+            // This can only be true if this column is a complex type.
+            // Complex types are expensive!
+            return maxScore;
+        }
+
+        checkArgument(filters.size() == 1);
+
+        Map.Entry<Subfield, TupleDomainFilter> filter = Iterables.getOnlyElement(filters.entrySet());
+
+        if(! filter.getKey().getPath().isEmpty()){
+            // The column is a complex type. Expensive!
+            return maxScore;
+        }
+
+        TupleDomainFilter tdf = filter.getValue();
+
+        if(tdf instanceof BigintRange) {
+            if(((BigintRange) tdf).isSingleValue()){
+                // An equality! Easy and good
```

Comment:
I'd make this comment similar to others: `// Integer equality. Generally cheap.`

Reply:
(y)